### PR TITLE
overhaul to better support charmhub's interface pages

### PR DESCRIPTION
**Pull request by simskij**
- Rename all `charm.yaml` to `interface.yaml` and move information about version, status and whether it's internal there to better support scraping by Charmhub.
- Add `interfaces` charm for hosting common charm libraries specific for the curated interfaces.
- Add CI validation for the interface.yaml files.
- Fix all interface definitions not adhering to the schema.
- Update contributing instructions

### CODE REVIEW

**Comment by PietroPasotti** [comment]
This is going to break https://github.com/canonical/pytest-interface-tester
we should coordinate release so as not to block all CIs

File: .interface-validator.py
Comment (PietroPasotti) [inline .interface-validator.py:42]: I'd do:

```python

errors = []
for file in files:
    try: 
        raw_interface = self.read_yaml(file)
    except yaml.errors.YAMLError:
        errors.append(f" - {file} contains invalid yaml")
    try: 
        Interface.validate(**raw_interface)
    except pydantic.ValidationError:
        errors.append(f" - {file} invalid interface schema: {e}")
        
if errors:
    exit("validation completed with errors: {'\n'.join(errors)}")
```

Comment (PietroPasotti) [reply .interface-validator.py:42]: this way you don't just get the first error in the stack, but you see all of them in case there's multiple

File: .interface-validator.py
Comment (PietroPasotti) [inline .interface-validator.py:32]:
```suggestion
        return config
```

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
LJTM 

(looks javascript to me)

File: interfaces/__template__/v0/interface.yaml
Comment (simskij) [reply interfaces/__template__/v0/interface.yaml:11]: Duplicating it mainly for use cases where it's used outside of this repo. For instance, as a validation schema **in** a charm at runtime, or by charmhub so they won't have to infer information from the repo structure.

File: utils/interface-validator.py
Comment (PietroPasotti) [inline utils/interface-validator.py:4]:
```suggestion
"""Schema validator for `interface.yaml` files."""
```

File: utils/interface-validator.py
Comment (PietroPasotti) [inline utils/interface-validator.py:19]: this is missing some optional fields that can be used to configure pytest-interface-tester. See https://github.com/canonical/pytest-interface-tester?tab=readme-ov-file#customizing-the-fixtures-address

File: utils/interface-validator.py
Comment (PietroPasotti) [inline utils/interface-validator.py:67]: either put this in a separate try block, or put all three checks in the same block and catch all exceptions one after the other

File: utils/interface-validator.py
Comment (PietroPasotti) [inline utils/interface-validator.py:15]: the enum member names should be uppercase, but I'm not sure if in this case it complicates your life too much?

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
final nits! 
LLJTM (looks less like javascript to me)

File: utils/interface-validator.py
Comment (ghislainbourgeois) [inline utils/interface-validator.py:1]:
```suggestion
# Copyright 2024 Canonical
```

**Review by ghislainbourgeois — COMMENTED** [review]
LGTM, Pietro already provided a couple of comments that I do agree with.

File: pyproject.toml
Comment (simskij) [reply pyproject.toml:9]: i share your understanding. however, we're not currently publishing interfaces on pip.

File: utils/interface-validator.py
Comment (PietroPasotti) [inline utils/interface-validator.py:24]: two blank lines between toplevel definitions. We should add a `fmt` tox env and a CI check to match

File: utils/interface-validator.py
Comment (PietroPasotti) [inline utils/interface-validator.py:16]: inconsistent newlines

File: utils/interface-validator.py
Comment (PietroPasotti) [inline utils/interface-validator.py:49]:
```suggestion
    """Error raised when the location of an interface.yaml spec file is inconsistent with its contents."""
```

File: utils/interface-validator.py
Comment (PietroPasotti) [inline utils/interface-validator.py:48]:
```suggestion
class ConsistencyError(Exception):
```

**Review by PietroPasotti — APPROVED** [review]
some final nits, doesn't need another round from me.
After some formatting we should be good to go.
GJ!

File: utils/interface-validator.py
Comment (gruyaume) [inline utils/interface-validator.py:59]: Add return typing

File: utils/interface-validator.py
Comment (gruyaume) [inline utils/interface-validator.py:62]: Add typing and return typing

**Review by gruyaume — APPROVED** [review]
Looks good, some nits around typing

File: pyproject.toml
Comment (sed-i) [reply pyproject.toml:9]: I meant there should be only one source to obtain them :)

File: CONTRIBUTING.md
Comment (tonyandrewmeyer) [inline CONTRIBUTING.md:11]:
```suggestion
  - a `schema.py` file containing [Pydantic](https://pydantic.dev/) models that specify the app and unit databag model for either side of the interface.
```

File: CONTRIBUTING.md
Comment (tonyandrewmeyer) [inline CONTRIBUTING.md:14]:
```suggestion
  - a `interface_tests` directory in which you can put Python files containing interface tests. [Read more about interface tests](./README_INTERFACE_TESTS.md)
```

File: CONTRIBUTING.md
Comment (tonyandrewmeyer) [inline CONTRIBUTING.md:15]:
```suggestion
- under `docs/`, the JSON schemas generated from the Pydantic schemas. You can use command `tox -e build-json-schemas` to generate them automatically. Do not edit those files manually.
```

File: README.md
Comment (tonyandrewmeyer) [inline README.md:21]:
```suggestion
[Contributing a new interface specification is a simple and lightweight process](./CONTRIBUTING.md).
```

File: README.md
Comment (tonyandrewmeyer) [inline README.md:25]:
```suggestion
For the time being, to see available interfaces, their statuses, and schemas, browse the [interfaces directory](./interfaces).
```

File: interfaces/__template__/v0/interface.yaml
Comment (tonyandrewmeyer) [inline interfaces/__template__/v0/interface.yaml:3]:
```suggestion
# When proposing a new interface, opt for underscores over dashes.
```

or

```suggestion
# When proposing a new interface, opt for underscored over dashed.
```

File: interfaces/__template__/v0/interface.yaml
Comment (tonyandrewmeyer) [inline interfaces/__template__/v0/interface.yaml:4]: It's so annoying to refer to metadata in docs these days :disappointed:. I've found it easiest to just say "metadata" and leave it to the reader to substitute in `charmcraft.yaml` or `metadata.yaml`.

```suggestion
# The name corresponds to the line with the arrow in the charm metadata snippet below:
```

File: interfaces/__template__/v0/interface.yaml
Comment (tonyandrewmeyer) [inline interfaces/__template__/v0/interface.yaml:15]:
```suggestion
# is not designed to be used by others, nor should it appear on Charmhub:
```

File: interfaces/__template__/v0/interface.yaml
Comment (tonyandrewmeyer) [inline interfaces/__template__/v0/interface.yaml:22]: [LEGEND.md](LEGEND.md) is deleted in this PR, so I assume this should point people elsewhere?

File: utils/interface-validator.py
Comment (tonyandrewmeyer) [inline utils/interface-validator.py:7]: * pyyaml isn't a dep in tox.ini, is it just being pulled in via something else (pydantic?)? It would be better if it was explicitly listed since it's explicitly imported.
* Does this repo have a formatter that includes imports? These would normally be:

```suggestion
import glob
import re

import yaml
```

ie. sections for stdlib and third-party, and in alphabetical order. But maybe that's not the style here?

Similar with the `from x import y` ones below.

**Review by tonyandrewmeyer — COMMENTED** [review]
A few small doc suggestions :)

File: utils/interface-validator.py
Comment (simskij) [reply utils/interface-validator.py:7]: > pyyaml isn't a dep in tox.ini, is it just being pulled in via something else (pydantic?)? It would be better if it was explicitly listed since it's explicitly imported.

it is :) https://github.com/canonical/charm-relation-interfaces/blob/c4145711cd57c1f34077a4852d9fd0570f1752aa/tox.ini#L88

**Comment by dimaqq** [comment]
@simskij a question has come up why the badges had disappeared back then.
Can you add a comment on that for posterity?

**Pull request merged**
